lib/rust/ensogl/component/grid-view/src/header.rs

```diff
@@ -0,0 +1,655 @@
+            Occupied(mut entry) => {
+                entry.get_mut().section_rows = rows;
+                entry.into_mut()
+            }
+            Vacant(lack_of_entry) => {
+                let new_entry = free_headers.pop().unwrap_or_else(create_new_entry);
+                self.grid.add_child(&new_entry);
+                let new_header_entry =
+                    VisibleHeader { section_rows: rows, entry: new_entry };
+                lack_of_entry.insert(new_header_entry)
+            }
+        };
+        let entry_frp = entry.entry.entry.frp();
+        entry_frp.set_model(model);
+        entry_frp.set_location((entry.section_rows.start, col));
+        let position = entry.header_position(col, entry_size, viewport);
+        entry.entry.set_position_xy(position);
+        (entry.section_rows.start, col, position)
+    }
+}
+
+
+
+// ================
+// === GridView ===
+// ================
+
+/// A template for [`GridView`] structure, wrapping any version of Grid View, and where entry
+/// parameters and model are separate generic arguments.
+#[derive(CloneRef, Debug, Derivative)]
+#[derivative(Clone(bound = ""))]
+pub struct GridViewTemplate<
+    Entry,
+    InnerGridView,
+    HeaderEntry,
+    HeaderModel: frp::node::Data,
+    HeaderParams,
+> {
+    frp:        Frp<HeaderModel>,
+    model:      Rc<Model<InnerGridView, HeaderEntry, HeaderParams>>,
+    entry_type: PhantomData<Entry>,
+}
+
+impl<Entry, InnerGridView, HeaderEntry, HeaderModel: frp::node::Data, HeaderParams> Deref
+    for GridViewTemplate<Entry, InnerGridView, HeaderEntry, HeaderModel, HeaderParams>
+{
+    type Target = InnerGridView;
+
+    fn deref(&self) -> &InnerGridView {
+        &self.model.grid
+    }
+}
+
+/// Grid View with Headers.
+///
+/// This is an extended version of [the base Grid View](crate::GridView). Here each column is
+/// assumed to be arranged in sections. The first entry of each section is treated as header. When
+/// the section is scrolled down, the header remains visible.
+///
+/// The structure derefs to the [base GridView FRP API](crate::Frp). To access the API specific
+/// to headers use [`Self::header_frp`] method.
+///
+/// # Headers
+///
+/// Those headers which are pushed down can (and should) be instantiated in a separate layer, passed
+/// with `set_layers` input. User should ensure, that the layer is displayed over the normal
+/// entries. Those headers can be the same type as main entries, but does not have to. They should
+/// have the same [`Entry::Params`] type, however.
+///
+/// ## Requesting for Models
+///
+/// Only the pushed-down, currently visible headers are instantiated. The models for the header
+/// instances and information about their section should be provided on demand, as a reaction
+/// for [`Frp::section_info_needed`] event, by emitting [`Frp::section_info`] with proper data.
+///
+/// **Important**. The [`Frp::section_info_needed`] are emitted once when needed and not repeated
```

Comment:
This explains the usage of `section_info_needed` but this should better still also be explained in the docs for the endpoint.